internal/transport/http2_client.go

```diff
@@ -1010,6 +1010,18 @@ func (t *http2Client) Close(err error) {
 	}
 	t.mu.Unlock()
 
+	// Append info about previous goaways if there were any, since this may be important
+	// for understanding the root cause for this connection to be closed.
+	_, goAwayDebugMessage := t.GetGoAwayReason()
```

Comment:
`GetGoAwayReason` is implemented as follows:

```golang
func (t *http2Client) GetGoAwayReason() (GoAwayReason, string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.goAwayReason, t.goAwayDebugMessage
}
```

This means that the instead of unlocking `t.mu` on line 1011, we can instead do the following:
```golang
// Don't unlock on line 1011
goAwayDebugMessage := t.goAwayDebugMessage
t.mu.Unlock()
...
```

Reply:
You're right, we could do that.

Reply:
wait, may be I am getting confused with line number but `t.controlBuf.put()` aquire a lock as well so we need to unlock before that. I think we need to make sure that we unlock in case of timeout as well

Reply:
@purnesh42H `t.controlBuf.put()` has a separate `mu` lock and is accessed with `t.controlBuf.mu` for control buffers resources, and here we are dealing with `t.mu` for htt2Client.

Reply:
but `outgoingGoAwayHandler` is acquiring lock on `t.mu` which is followed by `t.controlBuf.put()`. We need to just make sure we unlock the  `outgoingGoAwayHandler` lock in case of timeout as well

Reply:
exactly, please refer to this comment.
https://github.com/grpc/grpc-go/pull/7662#discussion_r1787209735